src/lib/yang/state.lua

```diff
@@ -0,0 +1,92 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+module(..., package.seeall)
+
+local lib = require("core.lib")
+local shm = require("core.shm")
+local xpath = require("lib.yang.xpath")
+local yang = require("lib.yang.yang")
+local yang_data = require("lib.yang.data")
+local counter = require("core.counter")
+
+local counter_directory = "/apps"
+
+local function find_counters(pid)
+    local path = shm.root.."/"..pid..counter_directory
+    local counters = {}
+    for _, c in pairs(lib.files_in_directory(path)) do
+        local counterdir = "/"..pid..counter_directory.."/"..c
+        counters[c] = shm.open_frame(counterdir)
+    end
+    return counters
+end
+
+function collect_state_leaves(schema)
```

Comment:
Here I think it would be a good idea to abstract this a bit: make a function that takes a schema and path and returns a state-getting function, and then we can call that function multiple times on a process with minimal overhead.  The yang modules will eventually be caching all of these schema->function calls.